Use jinja2 for templating

keszybz opened the pull request:
m4 use is dropped, and the two parallel substitution systems in meson are unified.

keszybz commented:
I guess the CI will fail with missing deps… This will need to be figured out.

bluca commented:
> I guess the CI will fail with missing deps… This will need to be figured out.

Pushed to Salsa for the bionic tests. PR for the CentOS tests: https://github.com/systemd/systemd-centos-ci/pull/382

bluca commented on .mkosi/mkosi.arch:
```diff
@@ -41,6 +41,7 @@ BuildPackages=
         pkgconfig
         python
         python-lxml
+        python-jinja2
```
looks like this is just `python-jinja` https://github.com/systemd/systemd-centos-ci/pull/382#discussion_r633137477

bluca commented:
Looks very nice! But the CI is a bit unhappy:

```
305/890 test-sysusers                           FAIL     0.02 s (exit status 127)

--- command ---
18:45:07 /usr/bin/env /home/runner/work/systemd/systemd/build/test/test-sysusers.sh /home/runner/work/systemd/systemd/build/systemd-sysusers
--- stderr ---
/usr/bin/env: ‘/home/runner/work/systemd/systemd/build/test/test-sysusers.sh’: No such file or directory
```

evverx commented:
@bluca judging by https://autopkgtest.ubuntu.com/results/autopkgtest-bionic-upstream-systemd-ci-systemd-ci/bionic/amd64/s/systemd-upstream/20210516_223707_2a255@/log.gz (from https://github.com/systemd/systemd/pull/19598)
```
dpkg-checkbuilddeps: error: Unmet build dependencies: python3-jinja2:native
dpkg-buildpackage: warning: build dependencies/conflicts unsatisfied; aborting
dpkg-buildpackage: warning: (Use -d flag to override.)
blame: https://salsa.debian.org/systemd-team/systemd.git#upstream-ci
badpkg: rules build failed with exit code 3
autopkgtest [22:36:55]: ERROR: erroneous package: rules build failed with exit code 3
Creating nova instance adt-bionic-amd64-systemd-upstream-20210516-223416 from image adt/ubuntu-bionic-amd64-server-20210516.img (UUID ef35f45b-f21e-48c4-a39c-82ebd5575ffa)... 
```
Ubuntu CI is broken

bluca commented:
> @bluca judging by https://autopkgtest.ubuntu.com/results/autopkgtest-bionic-upstream-systemd-ci-systemd-ci/bionic/amd64/s/systemd-upstream/20210516_223707_2a255@/log.gz (from #19598)
> 
> ```
> dpkg-checkbuilddeps: error: Unmet build dependencies: python3-jinja2:native
> dpkg-buildpackage: warning: build dependencies/conflicts unsatisfied; aborting
> dpkg-buildpackage: warning: (Use -d flag to override.)
> blame: https://salsa.debian.org/systemd-team/systemd.git#upstream-ci
> badpkg: rules build failed with exit code 3
> autopkgtest [22:36:55]: ERROR: erroneous package: rules build failed with exit code 3
> Creating nova instance adt-bionic-amd64-systemd-upstream-20210516-223416 from image adt/ubuntu-bionic-amd64-server-20210516.img (UUID ef35f45b-f21e-48c4-a39c-82ebd5575ffa)... 
> ```
> 
> Ubuntu CI is broken

Ah yes, jijnja is arch:all so the :native suffix is wrong - removed

keszybz commented:
Updated:
- no "2" for arch
- add `depends:...` in `test-sysusers` tests.

Let's see if this helps.

keszybz commented:
Updated:
- `env python3` not `env python`

bluca commented:
I'm not sure where semaphore is getting the list of build-dependencies - it's not from the upstream-ci branch in Salsa, as I've updated it. Maybe from the uploaded package - in which case python3-jinja2 needs to be added manually to .semaphore/semaphore-runner.sh for now

bluca commented:
This is a strange one, in CentOS:

```
tmpfiles.d/meson.build:35:8: ERROR:  Tried to create target "systemd.conf", but a target of that name already exists.
```

keszybz commented:
One more attempt:
- the oss-fuzz pr has been merged
- workaround for old meson
- python3-jinja2 is mentioned in `.semaphore/semaphore-runner.sh`

keszybz commented:
Another attempt:
- require bash explicitly where dash is not enough

poettering commented:
> m4 use is dropped, and the two parallel substitution systems in meson are unified.

Two parallel substitution systems? Can you explain?

I am not so at home in the Python world. Are you saying meson already use "jinja2" anyway? My quick RPM queries didn't show that? What am I missing?

I mean, i can't say I like m4 much, but just trying to understand the background of this change?

Hmm, is m4 a dep that is actually removed that way? isn#t it something pulled in indirectly anyway?

poettering commented on meson.build:
```diff
@@ -3741,7 +3746,7 @@ status = [
                                                               get_option('debug-tty')),
         'TTY GID:                           @0@'.format(tty_gid),
-        'users GID:                         @0@'.format(substs.get('USERS_GID')),
+        'users GID:                         @0@'.format(conf.get('USERS_GID')),
```
hmm? can you explain?

keszybz replied:
It's as @bluca wrote: we had two big `configuration_data` objects in the meson config. (There are in fact more, because I add one in this series, and there's one for efi… But those other have a handful over variables for specific purposes and don't matter). `conf` and `substs` were inherited from the original autotools system. In the past there was even a third set, but @yuwata removed it in 348b44372f36010d48d9a7dda14ef67155753a71. And the the two/three systems have very similar data, but with different variable names, because of historical reasons. This means we have duplicated work when changing things, and we need search for different variable names, etc.

poettering commented on units/systemd-nspawn@.service.in:
```diff
@@ -28,9 +28,9 @@ Delegate=yes
-# change them!
+{# Enforce a strict device policy, similar to the one nspawn configures when it
+ # allocates its own scope unit. Make sure to keep these policies in sync if you
+ # change them! #}
```
maybe add a broef comment here that explains the { thing, so that this is not dropped again

bluca commented:
> > m4 use is dropped, and the two parallel substitution systems in meson are unified.
> 
> Two parallel substitution systems? Can you explain?

I think it's meant to reference the two sets of conf (`configuration_data()`) objects we maintain in Meson, with somewhat-parallel-but-not-really content:

https://github.com/systemd/systemd/blob/main/meson.build#L19

poettering commented on src/login/71-seat.rules.in:
```diff
@@ -59,9 +59,11 @@ SUBSYSTEM=="pci", ATTRS{vendor}=="0x1ab8", ATTRS{device}=="0x4005", TAG+="seat",
 # the child if we notice that the parent wasn't recognized yet.
 
 # Match parent
+{% raw -%}
```
maybe add a comment right here explaining this

poettering commented on shell-completion/bash/systemctl.in:
```diff
@@ -21,6 +21,7 @@ __contains_word () {
     done
 }
 
+{% raw -%}
```
add comment why?

poettering commented on tmpfiles.d/meson.build:
```diff
@@ -33,7 +33,8 @@ in_files = ['etc.conf',
 foreach file : in_files
         custom_target(
-                file,
+                # XXX: workaround for old meson. Drop when upgrading.
```
maybe add version info to this comment? and a FIXME tag?

keszybz replied:
`XXX` is like `FIXME`, no? (And it's not really FIXME… It's something ugly but beyond our control for now. So I think XXX is better than FIXME).  I have no idea in which version exactly this was fixed, and I don't want to do meson git archaeology to find out. In the commit message I wrote that it reproduces with the oldest meson version I could easily install, 0.49.2 from buster.

poettering replied:
lgtm.com picks up FIXME as a magic word and tracks it as things to fix. (Doesn't coverity too?) Are you sure it also picks up XXX?

keszybz replied:
I don't think coverity does anything about either of the two.
Do we make any use of the tracking in lgtm.com? I have noticed the comments from coverity, and it could be useful to notice that a pull request has a fixme, if the fixme was intended to be fixed before the pull request is merged. Here, if lgtm made a comment on the pull request, it would be just noise.

poettering replied:
gnome-builder picks up both FIXME + XXX.

I did look at the lgtm FIXME list every now and then. It's moderately useful I'd say

poettering commented:
hmm, in the logind udev rules you add special "raw" escaping around %k lines. In rules.d/99-systemd.rules.in there's another line like that. shouldn't that get the same treatment? (and there might be more with other % chars?)

poettering commented:
btw, i wished meson would just make this decision for us and pick one templating engine and make it particularly well supported. Sounds weird, that everyone picks something different and needs to put glue scripts in place to make it work. Templating source code files seems like such a basic feat...

bluca commented:
On bionic-amd64:

```
====== test-sysusers.sh =======
/tmp/autopkgtest.sGZouq/build.WSD/systemd/debian/tests/root-unittests: 14: /tmp/autopkgtest.sGZouq/build.WSD/systemd/debian/tests/root-unittests: /usr/lib/systemd/tests/test-sysusers.sh: Permission denied
FAIL: test-sysusers.sh (code: 126)
```

keszybz commented:
> btw, i wished meson would just make this decision for us and pick one templating engine and make it particularly well supported. Sounds weird, that everyone picks something different and needs to put glue scripts in place to make it work. Templating source code files seems like such a basic feat...

meson already has a few templating engines, in some sense (see the `format` and `output_format` options in https://mesonbuild.com/Reference-manual.html#configure_file). But yeah, when working on this I thought it would be very nice if meson could provide direct support for jinja2 templates or something similar. But maybe it's not necessary, because there's many different ways to do this... Instead, it would be nice if meson allowed the configuration_data() object to be saved to a json, so it would be easy to read it back in and feed whatever template engine. This would make some things easier and more robust, and wouldn't be a lot of work for meson.

keszybz commented:
> hmm, in the logind udev rules you add special "raw" escaping around %k lines. In rules.d/99-systemd.rules.in there's another line like that. shouldn't that get the same treatment? (and there might be more with other % chars?)

The jinja2 syntax uses `{{`, `{#`, `{%`, `}}`, `#}`, `%}`. A percent sign in other places is not an issue.

keszybz commented:
> I am not so at home in the Python world. Are you saying meson already use "jinja2" anyway? My quick RPM queries didn't show that? What am I missing?

Sorry for being unclear. By "meson" I meant out meson config in systemd, not meson itself. Meson doesn't use jinja2 itself, afaik. 

> Hmm, is m4 a dep that is actually removed that way? isn#t it something pulled in indirectly anyway?

It is removed. It might be pulled in by something else, but not that many things require it. Automake and the rest of the autotools stack, bison+flex, redhat-lsb-core, some -devel packages. So yeah, it seems entirely reasonable to not have it.

keszybz commented:
Updated. I took a different approach for `test-sysusers.sh`: it now uses `configuration_file()` again, but with `conf` not `substs`. See the commit message for a lengthy justification. I hope this will allow `bionic-*` to pass.

I added a commit to describe the basics of the syntax in docs/HACKING and give some pointers to the relevant documentation. I also extended the commit messages with justifications.

keszybz commented:
bionic-* fails on 'import dataclasses' which also happens in other PRs… I'll investigate.

But I don't understand the CIFuzz failure:
```
...
 Setting up python3-jinja2 (2.8-1ubuntu0.1) ...
...
 Running command: /usr/local/bin/python3 -c import jinja2
--- stdout ---

--- stderr ---
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'jinja2'



meson.build:620:8: ERROR: Problem encountered: python3 jinja2 missing
+ exit 1
```

evverx commented:
> But I don't understand the CIFuzz failure:

On OSS-Fuzz (and CIFuzz) the latest version of Python is built from scratch in the base-builder image. `/usr/lib/python3/dist-packages` (where `python3-jinja2` in installed by `apt-get`) isn't included in `PYTHONPATH` there:
```
['', '/usr/local/lib/python38.zip', '/usr/local/lib/python3.8', '/usr/local/lib/python3.8/lib-dynload', '/usr/local/lib/python3.8/site-packages']
```
`pip3 install` instead of `apt-get install` should probably fix the issue.

bluca reviewed (approved):
This looks good to me, and it seems CI is going to pass now. Is it ready to be merged?

keszybz commented:
Yes, I think it's good to merge.

yuwata commented on src/timesync/meson.build:
```diff
@@ -34,15 +34,16 @@ libtimesyncd_core = static_library(
         include_directories : includes,
```
Typo in commit message: ninja2 -> jinja2

yuwata commented on test/run-unit-tests.py:
```diff
@@ -1,7 +1,6 @@
 #!/usr/bin/env python3
```
This commit is already merged.

yuwata commented:
Rebased and force-pushed. CIs were already green. Merging.
